<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="stephens999" />

<meta name="date" content="2018-06-28" />

<title>wSVD</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/workflowr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">wSVD</h1>
<h4 class="author"><em>stephens999</em></h4>
<h4 class="date"><em>2018-06-28</em></h4>

</div>


<p><strong>Last updated:</strong> 2018-07-14</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(12345)</code> </summary></p>
<p>The command <code>set.seed(12345)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/stephens999/misc/tree/6abd1067c5fa16c51e911cc3b319cc9e01e5d85e" target="_blank">6abd106</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.RData
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/ALStruct_cache/
    Ignored:    data/.Rhistory
    Ignored:    docs/figure/.DS_Store

Untracked files:
    Untracked:  .dropbox
    Untracked:  Icon
    Untracked:  _workflowr.yml
    Untracked:  analysis/GTEX-cogaps.Rmd
    Untracked:  analysis/SPCAvRP.rmd
    Untracked:  analysis/eQTL.perm.rand.pdf
    Untracked:  analysis/ieQTL.perm.rand.pdf
    Untracked:  analysis/mash_bhat_z.Rmd
    Untracked:  analysis/mash_ieqtl_permutations.Rmd
    Untracked:  analysis/sc_bimodal.Rmd
    Untracked:  analysis/susie_example.Rmd
    Untracked:  analysis/test_sparse.Rmd
    Untracked:  analysis/z.txt
    Untracked:  code/multivariate_testfuncs.R
    Untracked:  data/4matthew/
    Untracked:  data/4matthew2/
    Untracked:  data/E-MTAB-2805.processed.1/
    Untracked:  data/ENSG00000156738.Sim_Y2.RDS
    Untracked:  data/GDS5363_full.soft.gz
    Untracked:  data/GSE41265_allGenesTPM.txt
    Untracked:  data/Muscle_Skeletal.ACTN3.pm1Mb.RDS
    Untracked:  data/Thyroid.FMO2.pm1Mb.RDS
    Untracked:  data/bmass.HaemgenRBC2016.MAF01.Vs2.MergedDataSources.200kRanSubset.ChrBPMAFMarkerZScores.vs1.txt.gz
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.NewSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/bmass.HaemgenRBC2016.Vs2.PreviousSNPs.ZScores.hclust.vs1.txt
    Untracked:  data/finemap_data/fmo2.sim/b.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out2_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/dap_out_snp.txt
    Untracked:  data/finemap_data/fmo2.sim/data
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.config
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.k4.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.ld
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.snp
    Untracked:  data/finemap_data/fmo2.sim/fmo2.sim.z
    Untracked:  data/finemap_data/fmo2.sim/pos.txt
    Untracked:  data/logm.csv
    Untracked:  data/m.cd.RDS
    Untracked:  data/m.cdu.old.RDS
    Untracked:  data/m.new.cd.RDS
    Untracked:  data/m.old.cd.RDS
    Untracked:  data/mainbib.bib.old
    Untracked:  data/mat.csv
    Untracked:  data/mat.txt
    Untracked:  data/mat_new.csv
    Untracked:  data/paintor_data/
    Untracked:  data/temp.txt
    Untracked:  data/y.txt
    Untracked:  data/y_f.txt
    Untracked:  docs/figure/eigen.Rmd/
    Untracked:  docs/figure/fmo2.sim.Rmd/
    Untracked:  docs/figure/newVB.elbo.Rmd/
    Untracked:  docs/figure/rbc_zscore_mash2.Rmd/
    Untracked:  docs/figure/rbc_zscore_mash2_analysis.Rmd/
    Untracked:  docs/figure/rbc_zscores.Rmd/
    Untracked:  docs/trend_files/
    Untracked:  docs/z.txt
    Untracked:  explore_udi.R
    Untracked:  output/fit.varbvs.RDS
    Untracked:  output/glmnet.fit.RDS
    Untracked:  output/test.bv.txt
    Untracked:  output/test.gamma.txt
    Untracked:  output/test.hyp.txt
    Untracked:  output/test.log.txt
    Untracked:  output/test.param.txt
    Untracked:  output/test2.bv.txt
    Untracked:  output/test2.gamma.txt
    Untracked:  output/test2.hyp.txt
    Untracked:  output/test2.log.txt
    Untracked:  output/test2.param.txt
    Untracked:  output/test3.bv.txt
    Untracked:  output/test3.gamma.txt
    Untracked:  output/test3.hyp.txt
    Untracked:  output/test3.log.txt
    Untracked:  output/test3.param.txt
    Untracked:  output/test4.bv.txt
    Untracked:  output/test4.gamma.txt
    Untracked:  output/test4.hyp.txt
    Untracked:  output/test4.log.txt
    Untracked:  output/test4.param.txt
    Untracked:  output/test5.bv.txt
    Untracked:  output/test5.gamma.txt
    Untracked:  output/test5.hyp.txt
    Untracked:  output/test5.log.txt
    Untracked:  output/test5.param.txt

Unstaged changes:
    Modified:   analysis/_site.yml
    Deleted:    analysis/chunks.R
    Modified:   analysis/eigen.Rmd
    Modified:   analysis/fmo2.sim.Rmd
    Modified:   analysis/newVB.Rmd

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephens999/misc/blob/6abd1067c5fa16c51e911cc3b319cc9e01e5d85e/analysis/wSVD.Rmd" target="_blank">6abd106</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-14
</td>
<td style="text-align:left;">
workflowr::wflow_publish(“analysis/wSVD.Rmd”)
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/stephens999/misc/695703edb06015f9232e5a1caee99d98eece6ec9/docs/wSVD.html" target="_blank">695703e</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-14
</td>
<td style="text-align:left;">
Build site.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephens999/misc/blob/d007eadb07206308ff6ff589f1d864a72293ff23/analysis/wSVD.Rmd" target="_blank">d007ead</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-14
</td>
<td style="text-align:left;">
workflowr::wflow_publish(“analysis/wSVD.Rmd”)
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/stephens999/misc/e79677dbe01c36a4479e673faa70d969fa0bc598/docs/wSVD.html" target="_blank">e79677d</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-13
</td>
<td style="text-align:left;">
Build site.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/stephens999/misc/blob/ece36e85c419fc887b35e59261499d4dd8345eb0/analysis/wSVD.Rmd" target="_blank">ece36e8</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-13
</td>
<td style="text-align:left;">
workflowr::wflow_publish(“analysis/wSVD.Rmd”)
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Motivated by the desire to apply SVD and related methods to non-gaussian data (eg single cell data), I want to suggest investigating “weighted” versions of SVD that allows each observation to have its own measurement-error variance (in addition to any common variance). We already have this kind of idea in flash and mash, but methods like softImpute and svd are potentially more scalable, and so it would be nice to implement fast general versions of these.</p>
<p>The working model “rank k” model is <span class="math display">\[X = UDV&#39; + Z + E\]</span> where <span class="math inline">\(X\)</span>, <span class="math inline">\(Z\)</span> and <span class="math inline">\(E\)</span> are all <span class="math inline">\(n \times p\)</span> matrices, and <span class="math inline">\(D\)</span> is a <span class="math inline">\(k \times k\)</span> diagonal matrix. The elements of <span class="math inline">\(E\)</span> are iid <span class="math display">\[E_{ij} \sim N(0,\sigma^2=1/\tau)\]</span> and <span class="math display">\[Z_{ij} \sim N(0,s^2_{ij})\]</span> where <span class="math inline">\(s_{ij}\)</span> are known.</p>
<p>Note: in softImpute (alternating least squares; ALS version) they replace <span class="math inline">\(UDV&#39;\)</span> by <span class="math inline">\(AB&#39;\)</span>, but the basic idea is the same. Also in softImpute they introduce an L2 penalty, which is a nice feature to have, and which I think may not complicate things much here. (to be checked!)</p>
<p>Given <span class="math inline">\(Z\)</span> we note that the mle for <span class="math inline">\(U,D,V\)</span> is given by the SVD of (<span class="math inline">\(X-Z\)</span>). Following the usual EM idea, each iteration we can replace <span class="math inline">\(Z\)</span> with its expectation <span class="math inline">\(\bar{Z} = E(Z | U,D,V)\)</span> where <span class="math inline">\(U,D,V\)</span> are the current values of these parameters. Then the M step becomes running SVD on <span class="math inline">\(X-\bar{Z}\)</span>.</p>
<p>Given <span class="math inline">\(U,D,V\)</span> define residuals <span class="math inline">\(R= X-UDV\)</span>. Then from the model <span class="math inline">\(R_{ij} | Z \sim N(Z_{ij}, \sigma^2)\)</span>. Then from standard Bayesian analysis of Gaussians we have: <span class="math display">\[Z_{ij} | R \sim N(\mu_1,1/\tau_1)\]</span> where <span class="math display">\[\mu_1 = \tau/\tau_1 R_{ij}\]</span> <span class="math display">\[\tau_1 = \tau + 1/s_{ij}^2\]</span>.</p>
<p>In particular the conditional mean of <span class="math inline">\(Z\)</span> needed for EM is: <span class="math display">\[\bar{Z}_{ij}= \tau/\tau_1 R_{ij}\]</span>.</p>
<p>Note that in the special case <span class="math inline">\(s_{ij}=\Inf\)</span>, which is like <span class="math inline">\(X_{ij}\)</span> is “missing”, this gives <span class="math inline">\(\bar{Z}_{ij} = R_{ij}\)</span>, and when we plug that in to get a “new” value of <span class="math inline">\(R\)</span> we get <span class="math inline">\(R_{ij} = X_{ij}-\bar{Z}_{ij} = (UDV)_{ij}\)</span>. That is, each iteration</p>
<p>If we look in the softImpute code this is exactly what they use to deal with missing data. For example, line 49 of <code>simpute.als.R</code> is<br />
<code>xfill[xnas] = (U %*% (Dsq * t(V)))[xnas]</code>.</p>
</div>
<div id="idea" class="section level1">
<h1>Idea</h1>
<p>Basically my idea is that we should be able to modify the softImpute code by replacing this line (and similar lines involving xfill) with something based on the above derivation…One advantage of this is that softImpute already deals with ridge penalty, and is well documented and fast…</p>
<p>Alternatively we could just implement it ourselves as below without the ridge penalty…</p>
</div>
<div id="code" class="section level1">
<h1>Code</h1>
<p>I started coding an EM algorithm that imputes <span class="math inline">\(Z\)</span> each iteration. I haven’t tested it, so there may be bugs … but the objective seems to increase. This code may or may not be useful to build on.</p>
<pre class="r"><code>wSVD = function(x,s,k,niter=100,tau=NULL){
  if(is.null(tau)){ #for now just estimate tau by residual variance from fit of first k svd
    x.svd = svd(x,k,k)
    tau = 1/mean((x - x.svd$u %*% diag(x.svd$d[1:k]) %*% t(x.svd$v))^2) 
  }
  n = nrow(x)
  p = ncol(x)
  z = matrix(0,nrow=n,ncol=p)
  sigma2 = rep(0,niter)
  obj = rep(0,niter)
  for(i in 1:niter){
    x.svd = svd(x-z,k,k) # could maybe replace this with a faster method to get top k pcs?
    R = x - x.svd$u %*% diag(x.svd$d[1:k]) %*% t(x.svd$v)
    tau1 = tau + 1/s^2 
    z =  (tau/tau1)*R
    sigma2[i] = 1/tau
    obj[i] = sum(dnorm(R, 0, sqrt(s^2+(1/tau)), log=TRUE))
  }
  return(list(svd = x.svd,sigma2=sigma2,obj=obj))
}</code></pre>
<p>This example just runs it on constant <span class="math inline">\(s\)</span>, so it should match regular svd. I ran it with two different values of <span class="math inline">\(\tau\)</span>, but I don’t think <span class="math inline">\(\tau\)</span> affects the mle here…</p>
<pre class="r"><code>set.seed(1)
n = 100
p = 1000
s = matrix(1,nrow=n,ncol=p)
x = matrix(rnorm(n*p,0,s),nrow=n,ncol=p)
x.wsvd = wSVD(x,s,3,10,1)
plot(x.wsvd$obj)</code></pre>
<p><img src="figure/wSVD.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of unnamed-chunk-2-1.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/stephens999/misc/blob/695703edb06015f9232e5a1caee99d98eece6ec9/docs/figure/wSVD.Rmd/unnamed-chunk-2-1.png" target="_blank">695703e</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-14
</td>
</tr>
</tbody>
</table>
<p></details></p>
<pre class="r"><code>x.wsvd2 = wSVD(x,s,3,10,1e6)
plot(x.wsvd2$svd$u[,1],svd(x,3,3)$u[,1])</code></pre>
<p><img src="figure/wSVD.Rmd/unnamed-chunk-2-2.png" width="672" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of unnamed-chunk-2-2.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/stephens999/misc/blob/695703edb06015f9232e5a1caee99d98eece6ec9/docs/figure/wSVD.Rmd/unnamed-chunk-2-2.png" target="_blank">695703e</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-14
</td>
</tr>
</tbody>
</table>
<p></details></p>
<pre class="r"><code>plot(x.wsvd$svd$u[,1],svd(x,3,3)$u[,1])</code></pre>
<p><img src="figure/wSVD.Rmd/unnamed-chunk-2-3.png" width="672" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of unnamed-chunk-2-3.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/stephens999/misc/blob/695703edb06015f9232e5a1caee99d98eece6ec9/docs/figure/wSVD.Rmd/unnamed-chunk-2-3.png" target="_blank">695703e</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-14
</td>
</tr>
</tbody>
</table>
<p></details></p>
<p>Now this version is non-constant variance. At least the objective is increasing…</p>
<pre class="r"><code>s = matrix(rgamma(n*p,1,1),nrow=n,ncol=p)
x = matrix(rnorm(n*p,0,s),nrow=n,ncol=p)
x.wsvd = wSVD(x,s,30,100)
plot(x.wsvd$obj)</code></pre>
<p><img src="figure/wSVD.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<details> <summary><em>Expand here to see past versions of unnamed-chunk-3-1.png:</em></summary>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://github.com/stephens999/misc/blob/695703edb06015f9232e5a1caee99d98eece6ec9/docs/figure/wSVD.Rmd/unnamed-chunk-3-1.png" target="_blank">695703e</a>
</td>
<td style="text-align:left;">
stephens999
</td>
<td style="text-align:left;">
2018-07-14
</td>
</tr>
</tbody>
</table>
<p></details></p>
</div>
<div id="applying-to-poisson-data" class="section level1">
<h1>Applying to Poisson data</h1>
<p>Here is outline of how we might apply this to Poisson data.</p>
<p>Let <span class="math inline">\(Y_{ij} \sim Poi(\mu_{ij})\)</span> be observations. Let <span class="math inline">\(m_{ij}\)</span> be a “current fitted value” for <span class="math inline">\(\mu_{ij}\)</span>, for example it could be the current value of <span class="math inline">\(Y_{ij} - Z_{ij}\)</span> if <span class="math inline">\(Z_{ij}\)</span> where <span class="math inline">\(Z_{ij}\)</span> is the estimated “measurement error” in the above. Or for initialization <span class="math inline">\(m_{ij}\)</span> could be the mean of <span class="math inline">\(X_{i\cdot}\)</span> if <span class="math inline">\(j\)</span> indexes genes.</p>
<p>Basically the idea is to do a Taylor series expansion of the Poisson likelihood about <span class="math inline">\(m_{ij}\)</span>. This leads to - set <span class="math inline">\(X_{ij} = \log(m_{ij}) + (Y_{ij}-m_{ij})/m_{ij}\)</span> - set <span class="math inline">\(s^2_{ij} = 1/m_{ij}\)</span> which we could apply wSVD to. And then repeat…</p>
</div>
<div id="other-issues" class="section level1">
<h1>Other issues</h1>
<ul>
<li><p>I think there is something of a literature on wSVD, but not much implemented. Would be good to look further.</p></li>
<li><p>We would want to estimate the residual precision <span class="math inline">\(\tau\)</span> too. I believe there is a EM update for that based on computing expected squared residuals (which will involve second moment of <span class="math inline">\(Z_{ij}\)</span>, which is available).</p></li>
<li><p>It would be nice if we could estimate the ridge penalty in softImpute by maximum likelihood/variational approximation (as in flash/PEER) rather than having to do cross-validation.</p></li>
</ul>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.3.2 (2016-10-31)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X El Capitan 10.11.6

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] workflowr_1.0.1.9001 Rcpp_0.12.16         digest_0.6.15       
 [4] rprojroot_1.3-2      R.methodsS3_1.7.1    backports_1.1.2     
 [7] git2r_0.21.0         magrittr_1.5         evaluate_0.10.1     
[10] stringi_1.1.7        whisker_0.3-2        R.oo_1.22.0         
[13] R.utils_2.6.0        rmarkdown_1.9        tools_3.3.2         
[16] stringr_1.3.0        yaml_2.1.18          htmltools_0.3.6     
[19] knitr_1.20          </code></pre>
</div>
</div>

<hr>
<p>
    
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->
<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1.9001
</p>
<hr>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
